src/runtime_src/core/common/api/xrt_xclbin.cpp

```diff
@@ -38,126 +41,919 @@
 
 namespace xrt {
 
-// class xclbin_impl - class for xclbin objects
-//
-// Life time of xclbin are managed through shared pointers.
-// A buffer is freed when last references is released.
-class xclbin_impl
+/**
+ * class xclbin::mem_impl - xrt::xclbin::mem_impl provides the implementation
+ *
+ *   for a device memory object.
+ *
+ */
+class xclbin::mem_impl
 {
-protected:
-  std::vector<char> m_axlf;
-  const axlf* m_top;
+public:
+  mem_impl(const struct mem_data& memdata, const int32_t memidx);
 
-  void init_axlf_handle()
+  const std::string&
+  get_name() const{
+    return m_name;
+  }
+  
+  uint64_t
+  get_base_address() const
   {
-    const axlf* tmp = reinterpret_cast<const axlf*>(m_axlf.data());
-    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
-      throw std::runtime_error("Invalid xclbin");
-    m_top = tmp;
+    return m_base_address;
+  }
+  
+  uint64_t
+  get_size() const
+  {
+    return m_size;
   }
 
-public:
-  xclbin_impl(const std::vector<char>& data)
+  uint8_t
+  get_used() const
   {
-    m_axlf = data;
-    // Set pointer of type axlf*, and check magic string
-    init_axlf_handle();
+    return m_used;
   }
 
-  xclbin_impl(const std::string& filename)
+  uint8_t
```

Comment:
The type should be defined in its own enum class in xrt_xclbin.h in xrt::xclbin::mem.  Let not expose the underlying xclbin pod types.